File: leetcode/python/40.Combination Sum II.py
def Sum(candidates,target):
  result = []
  if candidates is None or len(candidates) == 0:
    return result
  current = []
  candidates.sort()
  result = combinationSum(candidates,target,0,current,result)
  return result

def combinationSum(candidates,target,j,curr,result):
  if target == 0:
    temp = curr
    result.append(temp)
    print(result)
    return 
  for i in range(j,len(candidates)):
    if target < candidates[i]:
      return
    if i==j or candidates[i] != candidates[i-1]:
      curr.append(candidates[i])
      combinationSum(candidates,target-candidates[i],i,curr,result)
      curr.pop()

Sum([2,3,4,7],7)
  

File: leetcode/kotlin/40.Combination Sum II.kt
import java.util.*

internal class Solution {
    fun combinationSum2(candidates: IntArray?, target: Int): List<List<Int>> {
        val result: MutableList<List<Int>> = ArrayList()
        if (candidates == null || candidates.isEmpty()) return result
        val current = ArrayList<Int>()
        Arrays.sort(candidates)
        combinationSum(candidates, target, 0, current, result)
        return result
    }

    private fun combinationSum(
        candidates: IntArray,
        target: Int,
        j: Int,
        curr: ArrayList<Int>,
        result: MutableList<List<Int>>
    ) {
        if (target == 0) {
            val temp = ArrayList(curr)
            result.add(temp)
            return
        }
        for (i in j until candidates.size) {
            if (target < candidates[i]) {
                return
            }
            if (i == j || candidates[i] != candidates[i - 1]) {
                curr.add(candidates[i])
                combinationSum(candidates, target - candidates[i], i + 1, curr, result)
                curr.removeAt(curr.size - 1)
            }
        }
    }
}